src/flp/types.rs

```diff
@@ -685,6 +720,60 @@ pub(crate) fn decode_result_vec<F: FftFriendlyFieldElement>(
     Ok(data.iter().map(|elem| F::Integer::from(*elem)).collect())
 }
 
+/// This evaluates range checks on a slice of field elements, using a ParallelSum gadget evaluating
+/// many multiplication gates.
+///
+/// # Arguments
+///
+/// * `gadget`: A `ParallelSumGadget<F, Mul<F>>` gadget, or a shim wrapping the same.
+/// * `input`: A slice of inputs. This calculation will check that all inputs were zero or one
+///   before secret sharing.
+/// * `joint_randomness`: A joint randomness value, used to compute a random linear combination of
+///   individual range checks.
+/// * `chunk_length`: How many multiplication gates per ParallelSum gadget. This must match what the
+///   gadget was constructed with.
+/// * `num_shares`: The number of shares that the inputs were secret shared into. This is needed to
+///   correct constant terms in the circuit.
+///
+/// # Returns
+///
+/// This returns (additive shares of) zero with high probability if all inputs were zero or one, or
+/// a non-zero value otherwise.
+pub(crate) fn parallel_sum_range_checks<F: FftFriendlyFieldElement>(
+    gadget: &mut Box<dyn Gadget<F>>,
+    input: &[F],
+    joint_randomness: F,
+    chunk_length: usize,
+    num_shares: usize,
+) -> Result<F, FlpError> {
+    let f_num_shares = F::from(F::valid_integer_try_from::<usize>(num_shares)?);
+    let num_shares_inverse = f_num_shares.inv();
+
+    let mut output = F::zero();
+    let mut r_power = joint_randomness;
+    let mut padded_chunk = vec![F::zero(); 2 * chunk_length];
+
+    for chunk in input.chunks(chunk_length) {
+        // Construct arguments for the Mul subcircuits.
+        for (input, args) in chunk.iter().zip(padded_chunk.chunks_exact_mut(2)) {
+            args[0] = r_power * *input;
+            args[1] = *input - num_shares_inverse;
+            r_power *= joint_randomness;
+        }
+        // If the chunk of the input is smaller than chunk_length, use zeros instead of measurement
+        // inputs for the remaining calls.
+        for args in padded_chunk[chunk.len() * 2..].chunks_exact_mut(2) {
+            args[0] = F::zero();
+            args[1] = -num_shares_inverse;
+            r_power *= joint_randomness;
```

Comment:
Seems like it's safe to skip this as long as the last chunk is indeed the partial chunk.
```suggestion
```